### PR TITLE
whatwg url connection string parsing

### PR DESCRIPTION
**Pull request by maximelkin**
tldr: stop using legacy url.parse api and use new URL() instead

Changes:
https://nodejs.org/api/url.html#url_percent_encoding_in_urls

### CODE REVIEW

**Comment by kibertoad** [comment]
Would it be possible to add test for this?

also are you still sure you don't want to be a maintainer :-p? You make a great dynamic duo with @briandamaged

**Comment by maximelkin** [comment]
>Would it be possible to add test for this?
Ok, next evening

>also are you still sure you don't want to be a maintainer :-p?
¯\_(ツ)_/¯

**Comment by maximelkin** [comment]
Some tests for this logic are in test/tape/parse-connection.js

**Comment by kibertoad** [comment]
@maximelkin If there are no complaints after people upgrade to 0.20.11 and we don't need to release a hotfix, I think we could bump version to 0.21.0 and drop Node 8 already. April is close.

**Comment by kibertoad** [comment]
Since there is a separate 0.20 branch for hotfixes, and master is for 1.0.0, we can merge this breaking change after CI is green and documentation is updated.

File: lib/util/parse-connection.js
Comment (TrySound) [inline lib/util/parse-connection.js:1]: Since node 10 URL is available globally like in browsers. This require can be removed.

File: lib/util/parse-connection.js
Comment (maximelkin) [inline lib/util/parse-connection.js:9]: protocol is always string, so && protocol && removed

Comment (kibertoad) [reply lib/util/parse-connection.js:9]: Here is the reason why it was made defensive: #3743
It was failing under certain circumstances.

File: lib/util/parse-connection.js
Comment (maximelkin) [reply lib/util/parse-connection.js]: done

**Comment by kibertoad** [comment]
@maximelkin Could you add notes about incompatible changes to `UPGRADING.md`?

File: test/tape/parse-connection.js
Comment (kibertoad) [inline test/tape/parse-connection.js:113]: we also need a test for sqlite Windows path, e. g. 'C:/node/database.sqlite'

**Review by kibertoad — CHANGES_REQUESTED** [review]
See comments

File: lib/util/parse-connection.js
Comment (maximelkin) [reply lib/util/parse-connection.js:9]: This check now not actual, because url.parse can return protocol null, and in URL case this is impossible

**Comment by kibertoad** [comment]
@maximelkin Tests are failing now

**Comment by maximelkin** [comment]
yea, i know, just searched for fix

**Review by kibertoad — APPROVED** [review]

File: test/tape/parse-connection.js
Comment (maximelkin) [reply test/tape/parse-connection.js]: done

**Comment by kibertoad** [comment]
Awesome, thanks!

**Comment by kibertoad** [comment]
@maximelkin Can you also update https://github.com/knex/knex/blob/master/UPGRADING.md with notes about changes in how connection string is interpreted?

**Comment by maximelkin** [comment]
https://nodejs.org/api/url.html#url_percent_encoding_in_urls
If you use something unusual for hostname/database name/auth parameters in url (not A-z, digits, dot, dash) - check how it would work.

**Comment by kibertoad** [comment]
That would work, if there is also an explicit mention that previously rules for legacy API applied, and now the WHATWG API do, with the URL reference that you gave.

**Pull request merged**
